### PR TITLE
arch/stack: get correct stack remain and limit dump size when sp is not within stack

### PR DESCRIPTION
**Pull request by Donny9**
## Summary
1. using running_task to get current tcb because this task maybe wrong in interrupt context.
2. Limit dump size when sp is not within stack. we need to limit this dump size because if the stack size of task is over syslog buffer, it will overwrite other dump message.
## Impact
N/A
## Testing
Vela Ci

### CODE REVIEW

**Comment by Donny9** [comment]
> 



> 



> @Donny9
> 
> > Limit dump size when sp is not within stack. we need to limit this dump size because if the stack size of task is over syslog buffer, it will overwrite other dump message.
> 
> This is a use case limitation. Please make it conditional or deal with it another way.

Hello, this is a common issue for dump super large stack? If sp isn't within stack,  this case usually not caused by stack overflow, if dump all stack, you will see that there are a lot of deadbeef(if color support) content in the lower stack, they are useless, so we need to rely on remain to ignore them(deadbeed).

This PR will set limit size to 8K,  so when this case happened, dump_stack function will dump stack from base+remain --- base+remain + 8K(if stack grow from high addr to low addr), this can allows user to see the effective information they need.

File: arch/arm/src/common/arm_assert.c [hunk 1]
```diff
@@ -411,6 +425,11 @@ static void arm_dumpstate(void)
   /* Dump the user stack */
 
   arm_dump_stack("User", sp,
+#ifdef CONFIG_STACK_COLORATION
+                 up_check_stack_remain(),
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/arm_assert.c:429]: let's rename do_stackcheck to arm_check_stack and call this function inside arm_dump_stack?

File: arch/xtensa/src/common/xtensa_dumpstate.c [hunk 2]
```diff
@@ -44,6 +44,12 @@
 
 #ifdef CONFIG_DEBUG_ALERT
 
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define STACKDUMP_SIZE 8192
```
Comment (xiaoxiang781216) [inline arch/xtensa/src/common/xtensa_dumpstate.c:51]: let's update the risc-v too

File: arch/arm/src/common/arm_checkstack.c [hunk 3]
```diff
@@ -209,12 +209,12 @@ ssize_t up_check_tcbstack_remain(struct tcb_s *tcb)
 
 size_t up_check_stack(void)
 {
-  return up_check_tcbstack(this_task());
+  return up_check_tcbstack(running_task());
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/arm_checkstack.c:212]: let's update arm64 too

File: arch/arm/src/common/arm_assert.c [hunk 4]
```diff
@@ -335,7 +337,7 @@ static int assert_tracecallback(struct usbtrace_s *trace, void *arg)
  * Name: arm_dump_stack
  ****************************************************************************/
 
-static void arm_dump_stack(const char *tag, uint32_t sp,
+static void arm_dump_stack(const char *tag, uint32_t sp, ssize_t remain,
```
[the remark this answers is not in the document]
Comment (Donny9) [reply arch/arm/src/common/arm_assert.c]: Done!

File: arch/arm/src/common/arm_assert.c (hunk 1 above)
Comment (Donny9) [reply arch/arm/src/common/arm_assert.c]: Done! patch 2

File: arch/arm/src/common/arm_checkstack.c (hunk 3 above)
Comment (Donny9) [reply arch/arm/src/common/arm_checkstack.c:212]: Done! patch 1

File: arch/arm/src/common/arm_internal.h [hunk 5]
```diff
@@ -440,6 +440,7 @@ void arm_usbuninitialize(void);
 
 /* Debug ********************************************************************/
 #ifdef CONFIG_STACK_COLORATION
+size_t arm_stackcheck(void *stackbase, size_t nbytes);
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/arm_internal.h:443]: arm_stack_check

File: arch/arm64/src/common/arm64_internal.h [hunk 6]
```diff
@@ -352,6 +352,7 @@ void arm64_usbuninitialize(void);
 /* Debug */
 
 #ifdef CONFIG_STACK_COLORATION
+size_t arm64_stackcheck(void *stackbase, size_t nbytes);
```
Comment (xiaoxiang781216) [inline arch/arm64/src/common/arm64_internal.h:355]: ditto

File: arch/Kconfig [hunk 7]
```diff
@@ -837,6 +837,14 @@ config ARCH_STACKDUMP
 	---help---
 		Enable to do stack dumps after assertions
 
+if ARCH_STACKDUMP
+
+config ARCH_STACKDUMP_LIMIT_LENGTH
```
Comment (xiaoxiang781216) [inline arch/Kconfig:842]: LIMIT_LENGTH->MAX_LENGTH

File: arch/arm/src/common/arm_assert.c [hunk 8]
```diff
@@ -353,9 +354,12 @@ static void arm_dump_stack(const char *tag, uint32_t sp,
     {
       _alert("ERROR: %s Stack pointer is not within the stack\n", tag);
 
+      remain = size - arm_stackcheck((FAR void *)(uintptr_t)base, size);
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/arm_assert.c:357]: need:
```
#ifdef CONFIG_STACK_COLORATION
```

File: arch/xtensa/src/common/xtensa_dumpstate.c [hunk 9]
```diff
@@ -395,6 +395,21 @@ void xtensa_dumpstate(void)
     }
   else
     {
+#ifdef CONFIG_STACK_COLORATION
+      uint32_t remain;
+
+      remain = ustacksize - xtensa_stackcheck((uintptr_t)ustackbase,
```
Comment (xiaoxiang781216) [inline arch/xtensa/src/common/xtensa_dumpstate.c:401]: where is risc-v?

File: arch/arm/src/common/arm_internal.h (hunk 5 above)
Comment (Donny9) [reply arch/arm/src/common/arm_internal.h]: Done!

File: arch/arm64/src/common/arm64_internal.h (hunk 6 above)
Comment (Donny9) [reply arch/arm64/src/common/arm64_internal.h]: Done!

File: arch/Kconfig (hunk 7 above)
Comment (Donny9) [reply arch/Kconfig]: Done!

File: arch/arm/src/common/arm_assert.c (hunk 8 above)
Comment (Donny9) [reply arch/arm/src/common/arm_assert.c]: Done!

File: arch/xtensa/src/common/xtensa_dumpstate.c (hunk 9 above)
Comment (Donny9) [reply arch/xtensa/src/common/xtensa_dumpstate.c]: Done!

File: arch/risc-v/src/common/riscv_assert.c [hunk 10]
```diff
@@ -307,6 +307,35 @@ static inline void riscv_showtasks(void)
...
+ * Name: riscv_dump_stack
+ ****************************************************************************/
+
+static void riscv_dump_stack(const char *tag, uintptr_t sp,
+                             uintptr_t base, uint32_t size, bool force)
+{
+  uintptr_t top = base + size;
+
+  _alert("%s Stack:\n", tag);
+  _alert("sp:     %08" PRIx32 "\n", sp);
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_assert.c:320]: PRIxPTR

File: arch/risc-v/src/common/riscv_assert.c [hunk 11]
```diff
@@ -307,6 +307,35 @@ static inline void riscv_showtasks(void)
...
+  _alert("  base: %08" PRIx32 "\n", base);
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_assert.c:321]: PRIxPTR

File: arch/arm/src/common/arm_assert.c [hunk 12]
```diff
@@ -352,10 +352,17 @@ static void arm_dump_stack(const char *tag, uint32_t sp,
   else
     {
       _alert("ERROR: %s Stack pointer is not within the stack\n", tag);
+#ifdef CONFIG_STACK_COLORATION
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/arm_assert.c:355]: move after line 358

File: arch/risc-v/src/common/riscv_assert.c [hunk 13]
```diff
@@ -327,11 +327,19 @@ static void riscv_dump_stack(const char *tag, uintptr_t sp,
     }
   else
     {
+#ifdef CONFIG_STACK_COLORATION
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_assert.c:330]: move after line 342

File: arch/xtensa/src/common/xtensa_dumpstate.c [hunk 14]
```diff
@@ -395,6 +395,15 @@ void xtensa_dumpstate(void)
...
+      uint32_t remain;
+
+      remain = ustacksize - xtensa_stack_check((uintptr_t)ustackbase,
+                                               ustacksize);
+      ustackbase += remain;
+      ustacksize -= remain;
+#endif
+
       _alert("ERROR: Stack pointer is not within allocated stack\n");
       xtensa_stackdump(ustackbase, ustackbase + ustacksize);
```
Comment (xiaoxiang781216) [inline arch/xtensa/src/common/xtensa_dumpstate.c:415]: should we add force to xtensa_stackdump and remove the code dup?

File: arch/Kconfig [hunk 15]
```diff
@@ -837,6 +837,14 @@ config ARCH_STACKDUMP
 	---help---
 		Enable to do stack dumps after assertions
 
+if ARCH_STACKDUMP
+
+config ARCH_STACKDUMP_MAX_LENGTH
+	int "The maximum length for dump stack on assertions"
```
Comment (xiaoxiang781216) [inline arch/Kconfig:843]: depends on ARCH_STACKDUMP and remove if/endif

Comment (Donny9) [reply arch/Kconfig:841]: Done.

File: arch/xtensa/src/common/xtensa_dumpstate.c (hunk 14 above)
Comment (Donny9) [reply arch/xtensa/src/common/xtensa_dumpstate.c]: Done

File: arch/risc-v/src/common/riscv_assert.c (hunk 13 above)
Comment (Donny9) [reply arch/risc-v/src/common/riscv_assert.c:334]: Done!

File: arch/arm/src/common/arm_assert.c (hunk 12 above)
Comment (Donny9) [reply arch/arm/src/common/arm_assert.c:357]: Done!

File: arch/risc-v/src/common/riscv_assert.c (hunk 11 above)
Comment (Donny9) [reply arch/risc-v/src/common/riscv_assert.c]: Done!

File: arch/risc-v/src/common/riscv_assert.c (hunk 10 above)
Comment (Donny9) [reply arch/risc-v/src/common/riscv_assert.c]: Done!

**Comment by Donny9** [comment]
image
@xiaoxiang781216 need to retrigger CI?

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@davids5 could you take a look?

**Review by davids5 — APPROVED** [review]
@Donny9 @xiaoxiang781216 - Thank you for the Knob!

**Pull request merged**